Suggested edits to "Get Started: A predictive modeling case study" (#125)
tidymodels/tidymodels.org: learn/work/calibration-splits/index.qmd

---
title: "Automatic calibration splits for resamples"
categories:
  - post-processing
  - calibration
  - rsample
type: learn-subsection
weight: 6
description: | 

>   Learn how tidymodels generates automatic calibration sets for post-processing.

Reviewer: ```suggestion
  Learn how tidymodels generates automatic calibration sets for postprocessing.
```

A few years ago we decided to not hyphenate this (or preprocessing)

Reviewer 2: Okay great, I've updated it elsewhere in the doc as well

toc: true

> toc-depth: 2

Reviewer 3: ```suggestion
toc-depth: 3
```

due to the length of the article i think upping the TOC depth is nice, otherwise we don't update the toc that fast

Reviewer 2: Nice, thank you!

include-after-body: ../../../resources.html
---

```{r}
#| label: "setup"
#| include: false
#| message: false
#| warning: false
source(here::here("common.R"))
```

```{r}
#| label: "load"
#| include: false
#| message: false
#| warning: false
library(tidymodels)

pkgs <- c("tidymodels", "rsample")

theme_set(theme_bw() + theme(legend.position = "top"))
```

While preprocessing is the transformation of the predictors prior to a model fit, post-processing is the transformation of the predictions after the model fit. This could be as straightforward as limiting predictions to a certain range of values to as complicated as transforming them based on a separate calibration model. 

> 

Reviewer 3: Below we are using the term **primary model** which [we just started using](https://github.com/search?q=org:tidymodels++%22primary+model%22&type=code). I like the term but i think it be nice to properly define it, in terms of pre/model/post diagram/terminology

Reviewer: I agree. I would add text to this paragraph after the first sentence and start the next paragraph with "An additional..."

> A calibration model is used to model the relationship between the predictions based on the primary model and the true outcomes. An additional model means an additional chance to accidentially overfit. So when working with calibration, this is crucial: we cannot use the same data to fit our calibration model as we use to assess the combination of primary and calibration model. Using the same data to fit the primary model and the calibration model means the predictions used to fit the calibration model are re-predictions of the same observations used to fit the primary model. Hence they are closer to the true values than predictions on new data would be and the calibration model doesn't have accurate information to estimate the right trends (so that they then can be removed).

Reviewer 3: ```suggestion
A calibration model is used to model the relationship between the predictions based on the primary model and the true outcomes. An additional model means an additional chance to Accidentally overfit. So when working with calibration, this is crucial: we cannot use the same data to fit our calibration model as we use to assess the combination of primary and calibration model. Using the same data to fit the primary model and the calibration model means the predictions used to fit the calibration model are re-predictions of the same observations used to fit the primary model. Hence they are closer to the true values than predictions on new data would be and the calibration model doesn't have accurate information to estimate the right trends (so that they then can be removed).
```

> rsample provides a collection of functions to make resamples for empirical validation of prediction models. So far, the assumption was that the prediction model is the only model that needs fitting, i.e., a resample consists of an analysis set and an assessment set.

Reviewer 3: Do we have a rsample doc page for analysis set and assessment set?

> rsample provides a collection of functions to make resamples for empirical validation of prediction models. So far, the assumption was that the prediction model is the only model that needs fitting, i.e., a resample consists of an analysis set and an assessment set.

Reviewer: ```suggestion
[rsample](https://rsample.tidymodels.org/) provides a collection of functions to make resamples for empirical validation of prediction models. So far, the assumption was that the prediction model is the only model that needs fitting, i.e., a resample consists of an analysis set and an assessment set.
```

Most of our other pages link to pkgdown sites when referencing the package name.

> If we include calibration into our workflow (bundeling preprocessing, (primary) model, and post-processing), we want an analysis set, a calibration set, and an assessment set. 

Reviewer 3: ```suggestion
If we include calibration into our workflow (bundling preprocessing, (primary) model, and post-processing), we want an analysis set, a calibration set, and an assessment set. 
```

Here we describe how we create these sets automatically when you resample or tune a workflow with calibration. Note that you currently cannot create manual calibration splits and supply them to the functions in tune, like `fit_resamples()` or `tune_grid()`. If you want to do this, leave us a [feature request](https://github.com/tidymodels/rsample/issues/new?template=feature_request.md).

## General principles

> First principle: we leave the assessment set purely for performance assessment and rather treat the analysis set as "the data to fit" - except now we split it to fit two different models instead of just one. Or in other words, we take the data for the calibration set from the analysis set, not from the assessment set. 

Reviewer 3: I'm writing most of these as I read, so disregard this one if it is answered later and i forget to remove this one.

Should we explain why we decide to split the analysis set instead of the assessment set. or at least mention that we will go over why later?

Reviewer 3: and nit pick, i think using some highlighting on "first principle" and "second principle" would be nice.

> First principle: we leave the assessment set purely for performance assessment and rather treat the analysis set as "the data to fit" - except now we split it to fit two different models instead of just one. Or in other words, we take the data for the calibration set from the analysis set, not from the assessment set. 

Reviewer: ```suggestion
**First principle**: we leave the assessment set purely for performance assessment and rather treat the analysis set as "the data to fit" - except now we split it to fit two different models instead of just one. Or in other words, we take the data for the calibration set from the analysis set, not from the assessment set. 
```

Reviewer: but I can never remember which markdown syntax we are using so maybe just one asterisk will produce bold ¯\\_(ツ)_/¯

If you compare a model with calibration to one without, and you use the same resamples, you are also using the same assessment sets.

> "Taking data from the analysis set" means splitting up the analysis set to end up with ... an analysis set and a calibration set. Now we have two sets called analysis set, that's confusing. If we need to distinguish them, we'll refer to them as "outer" and "inner" analysis set for "before" and "after" the split for a calibration set.

Reviewer: I found the notion of "two sets called analysis set" confusing. I think it should say that we "further split our initial analysis set into two partitions..." or something similar. I still like the inner/outer notation, but I think the process could be better worded here.

```{r}
#| label: calibration
#| echo: false
#| out.height: 350
#| fig.align: "center"

> knitr::include_graphics("images/analysis-calibration-assessment.jpg")

Reviewer 3: How are these images created? if there are source files I would like for them to be included in this folder. Otherwise they will be hard to tweak or reproduce later on

Reviewer: On Chrome, I get: 

<img width="1112" height="754" alt="image" src="https://github.com/user-attachments/assets/f39df02b-fc84-4e19-a1ab-9bd65c9c6bc5" />

```

It is important to note that including calibration requires more data than fitting a workflow without calibration. For small datasets, you might run into situations where fitting either model (primary model or calibration model) is not particularly stable - or you might not even be able to construct a calibration set properly.

> Second principle: we try to mimick the data splitting process used to create the (outer) analysis set and the assessment set.

Reviewer: Bold for "Second principle"

> And lastly, the third principle: if we cannot make a calibration split, we fall back onto an empty calibration set rather than erroring.

Reviewer: same

> 

Reviewer: I think that we should have a callout before the last sentence of this section that warns the users that they will need a sizable amount of data to do calibration effectively.

Let's look at v-fold cross-validation as an example.

> ### Example: v-fold cross-validation

Reviewer: For consistency, outside of syntax, we use upper case V when naming or describing this scheme.

For v-fold cross-validation, we split the data into `v` folds and create `v` resamples by always using one of the folds as the assessment set while combining the other folds into the analysis set. 

> Since we want to have a calibration split similar in spirit, we are splitting the (outer) analysis set into (inner) analysis set and calibration set with the same proportion of observations going into the analysis set: 1 - 1/v.

Reviewer 3: 1 - 1/v should either be styled as code or as latex, but don't care which on

Reviewer: LaTeX pleas

- `sliding_index()` works in much the same way, with the key difference being that what we are sliding over aren't the rows directly but rather an index. This is useful when dealing with irregular series of data: while the number of rows which fall into a window can vary, the window length is the same across all resamples.
- `sliding_period()` takes the idea of the `sliding_index()` function one step further and aggregates the index into periods before sliding across. This is useful for aggregating, e.g., daily data into months and then defining the analysis and assessment set in terms of months rather than the daily index directly.

### Row-based splitting

Let's start with the row-based splitting done by `sliding_window()`. We'll use a very small example dataset. This will make it easier to illustrate how the different subsets of the data are created but note that it is too small for real-world purposes. Let's use a data frame with 11 rows and say we want to use 5 for the analysis set, 3 for the assessment set, and leave a gap of 2 in between those two sets. We can make two such resamples from our data frame.

> ![](images/calibration-split-window.jpg)

Reviewer 3: could we do 4 for assessment set and a gap of 1?

right now there is very little air inside the assessment set with regards to the text

Reviewer 3: i'm realizing that this would be a huge undertaking

Or, in code:
```{r}
#| label: sliding_window_rset

df <- data.frame(x = 1:11)
r_set <- sliding_window(df, lookback = 4, assess_start = 3, assess_stop = 5)
r_set
```

Let's take that first resample and split it into an (inner) analysis set and a calibration set. 

```{r}
#| label: sliding_window_rsplit

r_split <- get_rsplit(r_set, 1)
analysis(r_split)
```

> To mimick the split into (outer) analysis and assessment set, we'll calculate the proportion of the analysis set and apply it to the observations available for fitting both the primary model and the calibration model, creating our (inner) analysis set. Since the remaining observations could be for the calibration model or a gap, we also calculate the proportion of the assessment set and use it to construct the calibration set. 

Reviewer: > return the (outer) analysis set as the (inner) analsysis set

Typo + I think this should be reworded. So, ultimately, there is no calibration set, and our original analysis set remains intact? If so maybe say that? The "return" wording is confusing.

```{r}
#| label: sliding_window_props

lookback <- 4
assess_start <- 3
assess_stop <- 5

# absolute sizes in the  outer split
length_window <- lookback + 1 + assess_stop
length_analysis_outer <- lookback + 1
length_assess <- assess_stop - assess_start + 1

# relative sizes
prop_analysis <- length_analysis_outer / length_window
prop_assess <- length_assess / length_window

# absolute sizes in the inner split
length_analysis_inner <- ceiling(prop_analysis * length_analysis_outer)
length_calibration <- ceiling(prop_assess * length_analysis_outer)

c(length_analysis_inner, length_calibration)
```

Calculating the length of the calibration set rather than the gap, together with rounding up when translating proportions to new lengths within the outer analysis set means that we prioritize allocating observations to the (inner) analysis and calibration set over allocating them to the gap. In this example here, this means that we are not leaving a gap between the analysis set and the calibration set.

> However, rounding up for both (inner) analysis and calibration set when we don't have a gap could mean we end up allocating more observations than we actually have. So in that case, we try to take from the calibration set if possible and thus prioritzing fitting the prediction model over the calibration model.

Reviewer 3: ```suggestion
However, rounding up for both (inner) analysis and calibration set when we don't have a gap could mean we end up allocating more observations than we actually have. So in that case, we try to take from the calibration set if possible and thus prioritizing fitting the prediction model over the calibration model.
```

```{r}
#| label: sliding_window_split_incomplete_3

r_split <- get_rsplit(r_set, 3)
split_cal <- internal_calibration_split(r_split, .get_split_args(r_set))
split_cal
```

If we can fit a complete (inner) analysis set in the available rows, we do so, even if incomplete sets are allowed.

```{r}
#| label: sliding_window_split_incomplete_5

r_split <- get_rsplit(r_set, 5)
split_cal <- internal_calibration_split(r_split, .get_split_args(r_set))
split_cal
```

### Index-based splitting

When working with irregular series, it can be helpful to define resamples based on sliding over an index, rather than the rows directly. That index often is a form of time but it does not have to be. Let's take our previous data frame and use `x` as the index, with a small modification. The index doesn't go straight from 1 to 11 but rather is missing an observation at the index value of 2. We use the same  `lookback`, `assess_start`, and `assess_stop`. The difference is that now these define the resamples on the _index_ rather than the _rows_ directly.

![](images/calibration-split-index.jpg)

> We still get two resamples, however, the analysis set contains only 4 rows because only those fall into the window defined by the index.

Reviewer 3: Would it be beneficial to add the missing value at 1 or 5 such that the analysis sets have different length?

Reviewer 3: or i guess, that isn't interesting at all. because then we do the same as the previous section

```{r}
#| label: sliding_index_rsplit_1

r_split <- get_rsplit(r_set, 1)
analysis(r_split)

split_cal <- internal_calibration_split(r_split, .get_split_args(r_set))
split_cal
analysis(split_cal)
calibration(split_cal)
```

#### When we can't make a calibration split

Abstracting away from observed rows to an index does pose two challenges for us though.

For the second resample, the theoretical window for the (outer) analysis set is 2-6. Because it fits within the boundaries of the observed rows (which is 1 to 11), we construct this set, even though we don't observe anything at an index of 2. 

Based on `lookback`, `assess_start`, and `assess_stop`, we calculate the same allocation for the (inner) analysis set and the calibration set as in the previous example, when working on the rows directly with `sliding_window()`: three for the inner analysis set and two for the calibration set. However, the dataset we want to split only covers the range [3, 6].

```{r}
#| label: sliding_index_rsplit_2

r_split <- get_rsplit(r_set, 2)
analysis(r_split)
```

> The sliding splits slide over _the data_, meaning they slide over observed values of the index and they slide only within the boundaries of the observed index values. So here, we can only slide within [3, 6] and thus cannot fit an inner analysis set of three and a calibration set of two into it. As established earlier, we fall back onto an empty calibration set in such a situation.

Reviewer 3: we mention a couple of times that we fall back. should we mention that we fall back with a warning?

```{r}
#| label: sliding_index_split_anchor

# missing 3 is the anchor for the inner split of the first resample
df <- data.frame(x = c(1:2, 4:11))
r_set <- sliding_index(
  df,
  index = x,
  lookback = 4,
  assess_start = 3,
  assess_stop = 5
)
r_split <- get_rsplit(r_set, 1)

internal_calibration_split(r_split, .get_split_args(r_set))
```

Sliding over the observed index values means that we cannot make a calibration split in two situations:

- We don't observe the index value at the lower boundary of the (outer) analysis set.
- We don't observe the index value relative to which the calibration split would be defined.

In these situations, we fall back onto an empty calibration set rather than erroring.

We expect this to not be relevant in practice too often, as chances of this happening are lower with more observations and we assume you are only attempting calibration with sufficiently large datasets.

### Period-based splitting

For the third sliding function in the family, we first aggregate the index to periods and then slide across those, e.g., aggregate daily data to monthly periods. 

![](images/calibration-split-period.jpg)

> The principle of how to contruct a calibration split on the (outer) analysis set remains the same. The challenges of abstracting away from the rows, as illustrated for sliding over observed instances of an index also remain. Here, we slide over observed periods. We observe a period, if we observe an index within that period. 

Reviewer 3: ```suggestion
The principle of how to construct a calibration split on the (outer) analysis set remains the same. The challenges of abstracting away from the rows, as illustrated for sliding over observed instances of an index also remain. Here, we slide over observed periods. We observe a period, if we observe an index within that period. 
```

Reviewer: >  We observe a period, if we observe an index within that period.

I'd reword here too. Maybe it is comma placement but it doesn't make immediate sense to me.

For `lookback = 0`, we remain unable to make a calibration split. However, it is possible to choose a smaller period when defining the resamples, e.g., using weeks instead of months when aggregating daily data. In that case, the calibration split is also defined in terms of weeks, rather than months, and thus failing to split a single month.

> Now that we've looked at the principles at play here on very small datasets, let's look at a dataset with a sufficient amount of observations to consider using calibration. The `Chicago` dataset has daily ridership numbers for over 15 years. Here we are constructing resamples which use 2 weeks worth of data for the assessment set, and roughly 15 years (52*15 weeks) for the analysis set. That's enough to use some of those observations for a calibration set!

Reviewer: 💯

chicago_assess <- assessment(chicago_r_split)
range(chicago_assess$date)
```

The calibration set consists of one week from the (outer) analysis set. 
```{r}
#| label: chicago_isplit

split_cal <- internal_calibration_split(
  chicago_r_split,
  .get_split_args(chicago_r_set)
)
chicago_cal <- calibration(split_cal)
range(chicago_cal$date)
```

## Splits for bootstrap samples

For bootstrap samples, our goal of avoiding overfitting means we need to refine our approach to making calibration splits a little further. 

If our (outer) analysis set is a bootstrap sample of the training data, it likely contains several replications of an observation. In the illustration below, the first row of the original data is sampled twice into the (outer) analysis set, the second row three times, and the fourth row also twice.

```{r}
#| label: bootstrap-resample
#| echo: false
#| out.width: 700
#| fig.align: "center"
knitr::include_graphics("images/bootstrap-resample.jpg")
```

We don't want those duplicates to be split up into the inner analysis and the calibration set. Simply creating another bootstrap resample from the (outer) analysis set makes it possible for this to happen. 

> In the example illustration below, the third row of the (outer) analysis set is labeled `2-1`, as it is the first intance of row 2 from the original data. The other two instances of this row in the (outer) analysis set are labeled `2-2` and `2-3`. In the example, the `2-1` row gets sampled into the (inner) analysis set twice, while `2-2` and `2-3` are not sampled and thus end up in the calibration set. This means that row 2 from the original data is now in both the (inner) analysis set and the calibration set.

Reviewer: ```suggestion
In the example illustration below, the third row of the (outer) analysis set is labeled `2-1`, as it is the first intance of row 2 from the original data. The other two instances of this row in the (outer) analysis set are labeled `2-2` and `2-3`. In the example, the `2-1` row gets sampled into the (inner) analysis set twice, while `2-2` and `2-3` are not sampled and thus end up in the calibration set. This means that row 2 from the original data is now in both the (inner) analysis set and the calibration set. This is inappropriate. 
```

```{r}
#| label: bootstrap-regular
#| echo: false
#| out.width: 700
#| fig.align: "center"
knitr::include_graphics("images/bootstrap-regular.jpg")
```

> To avoid this, we could do a grouped resampling with the row ID as the group. Then all replications of a row would end up in the same set, either (inner) analysis or calibration. However, this would mean that rows in the calibration set are potentially not unique, unlike the typical bootstrap OOB sample.

Reviewer: ```suggestion
To avoid this, we could do a grouped resampling with the row ID as the group. Then all replications of a row would end up in the same set, either (inner) analysis or calibration. However, this would mean that rows in the calibration set are potentially not unique, unlike the typical bootstrap out-of-bag (OOB) sample.
```

```{r}
#| label: bootstrap-grouped
#| echo: false
#| out.width: 700
#| fig.align: "center"
knitr::include_graphics("images/bootstrap-grouped.jpg")
```

To prevent both these potential issues, we create the calibration split by sampling the (inner) analysis set, with replacement, from the pool of unique rows in the (outer) analysis set. The calibration set then consists of the rows in the (outer) analysis set that were not sampled into the (inner) analysis set, like for a typicall OOB sample.

```{r}
#| label: bootstrap-modified
#| echo: false
#| out.width: 700
#| fig.align: "center"
knitr::include_graphics("images/bootstrap-modified.jpg")
```

## Summary

> When using calibration, we typically need generous amounts of data because we need separate data for each of these tasks: fit the primary model, fit the calibration model, and assess performance. 

Reviewer: ```suggestion
When using calibration, we typically require generous amounts of data because three partitions are necessary: one for fitting the primary model, another for fitting the calibration model, and separate data to assess performance. 
```

When tuning or resampling a workflow with calibration, tidymodels will automatically turn classic resamples (consisting of an analysis and an assessment set) into resamples which include a calibration set in addition to the analysis and assessment set. This is done according to the following principles: 

- The assessment set remains untouched, while the analysis set is split into an (inner) analysis set and a calibration set. 
- The inner calibration split mimicks the outer split (into analysis and assesment set) as closely as possible.
- If we can't make a calibration split based on these basic principles, we skip the calibration.

For sliding splits of ordered data, applying those principles is a bit more complex than for other types of splits as the outer split into analysis and assessment is already a bit more complex. We've laid out the details of this here for reference.

> For bootstrap splits, we don't directly split the (outer) analysis set but rather sample the (inner) analysis set from the unique rows in the (outer) analysis set to avoid data leakage between (inner) analysis and calibration set.

Reviewer: Were these to be two different sentences (based on the line break)?
